src/bitmap/multiops.rs

```diff
@@ -0,0 +1,455 @@
+    bitmaps: impl IntoIterator<Item = Result<RoaringBitmap, E>>,
+) -> Result<RoaringBitmap, E> {
+    let mut iter = bitmaps.into_iter();
+    let mut containers = match iter.next().transpose()? {
+        None => Vec::new(),
+        Some(v) => v.containers,
+    };
+
+    for bitmap in iter {
+        merge_container_owned(&mut containers, bitmap?.containers, BitXorAssign::bitxor_assign);
+    }
+
+    RetainMut::retain_mut(&mut containers, |container| {
+        container.ensure_correct_store();
+        container.len() > 0
+    });
+
+    Ok(RoaringBitmap { containers })
+}
+
+fn merge_container_owned(
+    lhs: &mut Vec<Container>,
+    rhs: Vec<Container>,
+    op: impl Fn(&mut Store, Store),
+) {
+    for mut rhs in rhs {
+        match lhs.binary_search_by_key(&rhs.key, |c| c.key) {
+            Err(loc) => lhs.insert(loc, rhs),
+            Ok(loc) => {
+                let lhs = &mut lhs[loc];
+                match (&lhs.store, &rhs.store) {
+                    (Store::Array(..), Store::Array(..)) => lhs.store = lhs.store.to_bitmap(),
+                    (Store::Array(..), Store::Bitmap(..)) => mem::swap(lhs, &mut rhs),
+                    _ => (),
+                };
+                op(&mut lhs.store, rhs.store);
+            }
+        }
+    }
+}
+
+#[inline]
+fn try_multi_or_ref<'a, E: 'a>(
+    bitmaps: impl IntoIterator<Item = Result<&'a RoaringBitmap, E>>,
+) -> Result<RoaringBitmap, E> {
+    // This algorithm operates on bitmaps. It must deal with arrays for which there are not (yet)
+    // any others with the same key.
+    //
+    //   1. Eager cloning would create useless intermediate values that might become bitmaps
+    //   2. Eager promoting forces disjoint containers to converted back to arrays at the end
+    //
+    // This strategy uses COW to lazily promote arrays to bitmaps as they are operated on.
+    // More memory efficient, negligible wall time difference benchmarks
+
+    // Phase 1. Borrow all the containers from the first element.
+    let mut iter = bitmaps.into_iter();
+    let mut start = collect_starting_elements(iter.by_ref())?;
+    let start_size = start.len();
+
+    start.sort_unstable_by_key(|bitmap| Reverse(bitmap.containers.len()));
+    let mut start = start.into_iter();
+    let mut containers = match start.next() {
+        Some(c) => {
+            let c: Vec<Cow<Container>> = c.containers.iter().map(Cow::Borrowed).collect();
+            if c.is_empty() {
+                // everything must be empty if the max is empty
+                start.by_ref().skip(start_size).next();
+            }
+            c
+        }
+        None => {
+            return Ok(RoaringBitmap::new());
+        }
+    };
+
+    // Phase 2: Operate on the remaining containers
+    for bitmap in start {
+        merge_container_ref(&mut containers, &bitmap.containers, |a, b| *a |= b);
+    }
+    for bitmap in iter {
+        merge_container_ref(&mut containers, &bitmap?.containers, |a, b| *a |= b);
+    }
```

Comment:
Is it possible to merge the iterator just to reduce the amount of logic to duplicate?

```suggestion
    for bitmap in start.into_iter().map(Ok).chain(iter) {
        merge_container_ref(&mut containers, &bitmap.containers, |a, b| *a |= b);
    }
```

Reply:
I've done it for all operations; the change seems to not be significant (less than 1%)